### ISSUE TITLE
Enable secure connections, including WebSocket, behind TLS-terminating reverse proxy

### ISSUE THREAD

**Issue by scottf** [nats-io/nats.java#1103]
### Proposed change

Enable secure connections, including WebSocket, behind TLS-terminating reverse proxy

### Use case

See conversation in https://github.com/nats-io/nats.java/pull/1041

**Comment by scottf** [comment]
@flex-seongbok @OpenX-Public Moving the conversation here.
I'm certain I don't understand a few things about how the reverse proxy works. The original PR was just as band-aid as far as I could see that maybe addressed one specific situation. 
I think the real change needs to be more thorough. Maybe part of the problem is that the proxy is not aware of the workflow of the NATS connection and protocol, just the higher level protocol of wss.
I would be happy to take a contribution that fully handles a reverse proxy for both normal tls and wss

**Comment by flex-seongbok** [comment]
First, thank you for your attention. I hope this conversation makes NATS more perfect.
I think the point of this issue is client validation for cilent-server TLS configuration mismatch.
What is the purpose of this validation? There could be middlewares between client and server but this validation assume client and NATS server are connected directly.
The only important thing is upgrade to TLS but who did it is not important to client.
It could be TLS reverse proxy in front of plain NATS or TLS NATS itself or even plain reverse proxy in front of TLS NATS.
I suggests removing client validation below.
```
if (isTLSRequired && !serverInfo.isTLSRequired()) {
  throw new IOException("SSL connection wanted by client.");
}
else if (!isTLSRequired && serverInfo.isTLSRequired()) {
  throw new IOException("SSL required by server.");
}
```

It's midnight here so reply could be late.

**Comment by scottf** [comment]
So I've been talking to some other client developers. Is there any way you can just have your proxy be a pass through. The main problem is the mismatch here. Given this code, I'm open to suggestions. Any solution has to be optional and not the default. There could be another connection Option. There could be a custom data port implementation that might skip the upgrade to secure, which would affect either handshake/tls first or not.

```
void upgradeToSecureIfNeeded(NatsUri nuri) throws IOException {
    Options clientOptions = getOptions();
    if (clientOptions.isTlsFirst()) {
        this.dataPort.upgradeToSecure();
    }
    else {
        ServerInfo serverInfo = getInfo();
        boolean before2_9_19 = serverInfo.isOlderThanVersion("2.9.19");

        boolean isTLSRequired = clientOptions.isTLSRequired();
        boolean upgradeRequired = isTLSRequired;
        if (isTLSRequired && nuri.isWebsocket()) {
            // We are already communicating over "https" websocket, so
            // do NOT try to upgrade to secure.
            if (before2_9_19) {
                isTLSRequired = false;
            }
            upgradeRequired = false;
        }
        if (isTLSRequired && !serverInfo.isTLSRequired()) {
            throw new IOException("SSL connection wanted by client.");
        }
        else if (!isTLSRequired && serverInfo.isTLSRequired()) {
            throw new IOException("SSL required by server.");
        }
        if (upgradeRequired) {
            this.dataPort.upgradeToSecure();
        }
    }
}
```

**Comment by flex-seongbok** [comment]
Changing proxy to bypass is not an option especially modern network security trend moving towards to extracting TLS responsibility out of server.
Use TLS-terminating reverse proxy between Public and Private network so servers in Private network can free from TLS management.
If it needs TLS security between servers in Private network I would use Istio proxy that uses mTLS to protect server.

And yes we can introduce new option like 'allowTlsMissmatch' to solve this problem optionally but I think this problem is not an option.
it's problem of who's responsibility to check network protocol.
And I think each client or server is responsible for own side, not both.
On the client, it matters whether if the client TLS required and upgrade successful, not whether the server is configured to TLS required.
On the server, it matters whether if the server TLS required and reject plain request, not whether the client is configured to TLS required.

**Comment by flex-seongbok** [comment]
Another option is add secure websocket condition in tlsFirst
As mentioned below it's also true for tlsFirst
> // We are already communicating over "https" websocket, so
> // do NOT try to upgrade to secure.

**Comment by scottf** [comment]
I've asked for official comment on modifying this behavior. Currently we do not support or recommend proxy behavior that has mismatched TLS between client and server, as the code shows. If we support this there will be official behavior that is documented and implemented.

**Comment by OpenX-Public** [comment]
@scottf Hi.
When will this be fixed? We can't use the java-nats library until this is fixed, and I'm worried that other libraries might have the same problem.

**Comment by scottf** [comment]
> @scottf Hi. When will this be fixed? We can't use the java-nats library until this is fixed, and I'm worried that other libraries might have the same problem.

There is no timeframe for this. As I already said, I have raised the issue internally and it is being discussed. We do not consider this "broken". The current stance is that we discourage proxies to begin with and especially those that are mismatched.

**Comment by scottf** [comment]
We are in no way saying that your use case does not make sense nowadays. When it comes to NATS, where the server and client are already built to be secure and to load balance, that the addition of a proxy on top of a messaging app also adds latency.

Also, please let me re-iterate. I have taken your request seriously, I have raised the issue internally, for instance I brought it up at our client and server developers meeting today. This is really out of my hands.

Since the client is open source, you are welcome to modify it for your own use. Also, another suggestion I received at the meeting today was that your proxy could re-write the INFO statement's json that comes from the server, in order to simulate that the server is in secure mode.

**Comment by OpenX-Public** [comment]
Hi, please let me know if you have any updates.

**Comment by scottf** [comment]
@OpenX-Public @flex-seongbok It has been approved to put in an option to ignore the mismatch. It will be some boolean setting in Options that give the developer control. I will work on it this week. Thanks for your patience.

**Comment by OpenX-Public** [comment]
Okay, thanks!

**Comment by OpenX-Public** [comment]
Hi. We'd appreciate it if you could take care of that as well.

**Comment by scottf** [comment]
There is a solution coming. The root cause here is that the client does not like the mismatch where the client is secure but the server is not. There is a feature in the server called `allow_non_tls` which is different than when tls is required. But the java client is not aware of this setting. So instead of making a new option that is ignore the mismatch, I'm adding support for `allow_non_tls`

On the server side, you will need to configure the server like so:

```
tls {}
allow_non_tls = true
```

This will produce a flag in the server INFO that the Java fix will recognize. When the client see's this, it will ignore the mismatch, which is what I was trying to accomplish.

So to recap. 1) Fix coming. 2) Requires minor change to server config. 3) Requires proper support in Java client.

**Comment by scottf** [comment]
@OpenX-Public @flex-seongbok Did either of you try with the TLS first client options. I've been going over that code and the upgrade happens without the mismatch check. I'm in the process of building some unit tests to demonstrate.

**Comment by OpenX-Public** [comment]
@scottf 
I have alreay tried tls first option in java client. But it does not work. (TLS connection failed).

**Comment by scottf** [comment]
I wonder if there is another issue. If you look at the tls first code, it never does the mismatch check.

**Comment by flex-seongbok** [comment]
@scottf In my case I tested on https websocket. It's communicating over https but tlsFirst always do upgrade even it's already using TLS. I'm not sure tcp connection.

**Comment by OpenX-Public** [comment]
Hi, i found some document about TLS terminating reverse proxy.

Please see this and scroll down.

https://docs.nats.io/running-a-nats-service/configuration/securing_nats/tls

Will that solution solve the problem?

**Comment by scottf** [comment]
Take a look at this: [PR 1127 Line 587](https://github.com/nats-io/nats.java/pull/1127/files#diff-bbcf614ee7ee4ccd15b22d1f95fa40e4e642098386b6d1d14127dab5968d89feR587)

I made sure that there is no upgrade on websocket even in tls first, because the websocket code is already upgrade. So tls first was the wrong idea.

My current guess is that either client's websocket code simply does not work with your proxy or ssl context is not configured correctly. It would be useful to see what exception you get during connection. To do this, you might need a custom error listener to print the stack track of the exception in the "exceptionOccurred" method

**Comment by flex-seongbok** [comment]
@scottf 
This is the error stacktrace of tlsFirst option on wss websocket request over TLS-terminating proxy
Using io.nats:jnats:2.17.2, io.nats:nats-spring:0.5.7

```
Caused by: javax.net.ssl.SSLException: Unsupported or unrecognized SSL message
	at java.base/sun.security.ssl.SSLSocketInputRecord.handleUnknownRecord(SSLSocketInputRecord.java:451)
	at java.base/sun.security.ssl.SSLSocketInputRecord.decode(SSLSocketInputRecord.java:175)
	at java.base/sun.security.ssl.SSLTransport.decode(SSLTransport.java:111)
	at java.base/sun.security.ssl.SSLSocketImpl.decode(SSLSocketImpl.java:1505)
	at java.base/sun.security.ssl.SSLSocketImpl.readHandshakeRecord(SSLSocketImpl.java:1420)
	at java.base/sun.security.ssl.SSLSocketImpl.startHandshake(SSLSocketImpl.java:455)
	at java.base/sun.security.ssl.SSLSocketImpl.startHandshake(SSLSocketImpl.java:426)
	at io.nats.client.impl.SocketDataPort.upgradeToSecure(SocketDataPort.java:123)
	at io.nats.client.impl.NatsConnection.upgradeToSecureIfNeeded(NatsConnection.java:556)
	at io.nats.client.impl.NatsConnection.lambda$tryToConnect$2(NatsConnection.java:441)
	at java.base/java.util.concurrent.FutureTask.run$$$capture(FutureTask.java:264)
	at java.base/java.util.concurrent.FutureTask.run(FutureTask.java)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1136)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:635)
	at java.base/java.lang.Thread.run(Thread.java:833)
```

**Comment by flex-seongbok** [comment]
2.17.7-SNAPSHOT works fine connecting wss:// connection over TLS-terminating proxy without problem.
Thank you.

But I can't test tcp connection because my local to server network only supports http protocol not tcp. sorry.

**Comment by scottf** [comment]
> 2.17.7-SNAPSHOT works fine connecting wss:// connection over TLS-terminating proxy without problem. Thank you.
That's great. I really appreciate your patience and working this through with me. 2.17.7 will be release in the next day or so.

**Comment by OpenX-Public** [comment]
@scottf 

Hello.

Can you please guide me with an official document which options should be enabled in nats-server (Server) and which options should be enabled in nats-java (Client) for Websocket connection via WSS on TLS terminated reverse proxy?

**Comment by scottf** [comment]
It's not documented yet, it's just added/figured out. Client you should have to do nothing and server should have this. I thought you said it was working?
```
tls {}
allow_non_tls: true
```
